sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadViaSDF.java

```diff
@@ -0,0 +1,861 @@
+        } catch (Exception anyException) {
+          LOG.warn("Failed to close offset consumer for {}", topicPartition);
+        }
+      }
+
+      @Override
+      public long estimate() {
+        return memorizedBacklog.get();
+      }
+    }
+
+    @GetInitialRestriction
+    public OffsetRange initialRestriction(@Element KafkaSourceDescription kafkaSourceDescription) {
+      Map<String, Object> updatedConsumerConfig =
+          overrideBootstrapServersConfig(consumerConfig, kafkaSourceDescription);
+      try (Consumer<byte[], byte[]> offsetConsumer =
+          consumerFactoryFn.apply(
+              KafkaIOUtils.getOffsetConsumerConfig(
+                  "initialOffset", offsetConsumerConfig, updatedConsumerConfig))) {
+        consumerSpEL.evaluateAssign(
+            offsetConsumer, ImmutableList.of(kafkaSourceDescription.getTopicPartition()));
+        long startOffset;
+        if (kafkaSourceDescription.getStartReadOffset() != null) {
+          startOffset = kafkaSourceDescription.getStartReadOffset();
+        } else if (kafkaSourceDescription.getStartReadTime() != null) {
+          startOffset =
+              consumerSpEL.offsetForTime(
+                  offsetConsumer,
+                  kafkaSourceDescription.getTopicPartition(),
+                  kafkaSourceDescription.getStartReadTime());
+        } else {
+          startOffset = offsetConsumer.position(kafkaSourceDescription.getTopicPartition());
+        }
+        return new OffsetRange(startOffset, Long.MAX_VALUE);
+      }
+    }
+
+    @GetInitialWatermarkEstimatorState
+    public Instant getInitialWatermarkEstimatorState(@Timestamp Instant currentElementTimestamp) {
+      return currentElementTimestamp;
+    }
+
+    @NewWatermarkEstimator
+    public WatermarkEstimatorT newWatermarkEstimator(
+        @WatermarkEstimatorState Instant watermarkEstimatorState) {
+      return createWatermarkEstimatorFn.apply(watermarkEstimatorState);
+    }
+
+    @GetSize
+    public double getSize(
+        @Element KafkaSourceDescription kafkaSourceDescription,
+        @Restriction OffsetRange offsetRange)
+        throws Exception {
+      double numOfRecords =
+          ((HasProgress) restrictionTracker(kafkaSourceDescription, offsetRange))
+              .getProgress()
+              .getWorkRemaining();
+
+      // Before processing elements, we don't have a good estimated size of records and offset gap.
+      if (avgOffsetGap.containsKey(kafkaSourceDescription.getTopicPartition())) {
+        numOfRecords =
+            numOfRecords / (1 + avgOffsetGap.get(kafkaSourceDescription.getTopicPartition()).get());
+      }
+      return (!avgRecordSize.containsKey(kafkaSourceDescription.getTopicPartition())
+              ? 1
+              : avgRecordSize.get(kafkaSourceDescription.getTopicPartition()).get())
+          * numOfRecords;
+    }
+
+    @NewTracker
+    public RestrictionTracker<OffsetRange, Long> restrictionTracker(
```

Comment:
This method should be able to be GrowableOffsetRangeTracker. This will allow you to avoid the HasProgress cast in getSize